main/main.go

```diff
@@ -212,9 +212,17 @@ func runRelayer(logger logging.Logger,
 ) {
 	logger.Info(
 		"Creating relayer",
-		zap.String("blockchainID", sourceSubnetInfo.BlockchainID),
+		zap.String("originBlockchainID", sourceSubnetInfo.BlockchainID),
 	)
 
+	// Marks when the relayer has finished the catch-up process on startup.
+	// Until that time, we do not know the order in which messages are processed,
+	// since the catch-up process occurs concurrently with normal message processing
+	// via the subscriber's Subscribe method. As a result, we cannot safely write the
+	// latest processed block to the database without risking missing a block in a fault
+	// scenario.
+	doneCatchingUp := make(chan bool, 1)
```

Comment:
I think using a channel here is a little overengineered. We can make `doneCatchingUp` an atomic bool that is a field in the `Relayer` object that is initialized to `false`. 

We can then pass a pointer to that atomic bool to `sub.ProcessFromHeight`, which it can set once it finishes. Then we don't need the extra case in the select statement below.

`relayer.RelayMessage` can then just check the bool itself instead of it being passed in.

Reply:
So I repurposed the channel so that it could be used to pass an error to select statement in `ProcessLogs`. We'd probably want a channel to propagate errors anyway.

Reply:
Sorry this is a novel... We can chat offline if any of this is confusing. 

I think we have a few problems currently: 

1. Errors in the catchup goroutine don't actually get propogated up. We log the error in the goroutine, and then the main thread propogates a new error, instead of one error all the way up.
2. We need to worry about the catchup channel, instead of using idiomatic error handling.
3. Cancellation can only happen in one direction. An error in the catchup block will cause the main goroutine to cancel, but an error in the main goroutine, or context cancellation will leak the catchup goroutine. (We should also fix this with the `forwardLogs()` goroutine)
4. We should update the database when processing the historical blocks, otherwise if we crash/exit before completing, we lose all progress.
5. The current implementation will make it hard to run the catchup goroutine again after a disconnection.

I think we can solve these fairly easily with something approximating the following steps:

- Move the call to ` go s.forwardLogs()` out of `dialAndSubscribe`
- Spawn a goroutine that takes a **_child_** context and calls (if required) `ProcessFromHeight(getLatestProcessedBlockHeight())` and then once that's finished, it can call `forwardLogs`. Both of these functions should check if the context is still live periodically.
  - The child context is to that `ProcessLogs` can cancel it in the case of a disconnection. This way we can create a new instance of the goroutine to process any missed blocks while we were disconnected.
  - Alternatively, we could split the `ProcessFromHeight` and `forwardLogs` into separate goroutines contained by an errGroup, but `ProcessLogs` needs a way of knowing it's allowed to write to the db for logs coming from `ProcessFromHeight`. This means we could process the historical and new blocks at the same time, but at the cost of adding complexity.
- Now when we hit `case err := <-r.Subscriber.Err():`, we should cancel the child context, resubscribe, and then spawn a new goroutine with a new child context.
- Any errors from the goroutine containing `ProcessFromHeight` should probably not propogate all the way up. We should probably swallow and log them in `ProcessLogs` and then try to resubscribe.

Reply:
I want to avoid overengineering a non-critical feature. Zooming out to focus on the functional issues, I think the main feature the current implementation is missing is the fact that if the relayer fails during catch up, we lose all progress. As I mentioned in another comment, we don't lose all progress, since the relayer checks if a message has already been delivered before aggregating signatures, but we can certainly do better.

A simple extension to the existing implementation would be to add additional data to the `WarpMessageInfo` to denote if it's a catch-up message. If it is, then we'd always write it to the DB. Otherwise, we only write to the DB if the catch up process is finished. That way, we can process catch up messages and live messages concurrently on startup, and also checkpoint our progress.

I'm open to the refactors you've described, but would rather implement them in a standalone PR (if we decide they're necessary) rather than expanding the scope of this PR.

Reply:
> 1. Errors in the catchup goroutine don't actually get propogated up. We log the error in the goroutine, and then the main thread propogates a new error, instead of one error all the way up.

While true, the main goroutine isn't distinguishing between error conditions from the catch up goroutine, so the boolean suffices.

> 2. We need to worry about the catchup channel, instead of using idiomatic error handling.

An error channel is idiomatic for passing errors from child goroutine to parent. We could have separate error/result goroutines, but for the reasons mentioned in my reply to 1., we can get away with just a single channel for now.

> 3. Cancellation can only happen in one direction. An error in the catchup block will cause the main goroutine to cancel, but an error in the main goroutine, or context cancellation will leak the catchup goroutine. (We should also fix this with the forwardLogs() goroutine)

If the main goroutine exits, all child goroutines will exit, so leaky goroutines aren't an issue in practice.

> 4. We should update the database when processing the historical blocks, otherwise if we crash/exit before completing, we lose all progress.

Agreed, it would be better if this were not the case

> 5. The current implementation will make it hard to run the catchup goroutine again after a disconnection.

The catch-up routine gets logs via RPC not websocket, so we don't need a persistent connection for correctness.

Reply:
> While true, the main goroutine isn't distinguishing between error conditions from the catch up goroutine, so the boolean suffices.

What I'm getting at here is that the error logs become uncorrelated and might be difficult to find the root cause of an error in the logs. If we at least put the error into the channel instead of a bool, we can pass that error up.

> An error channel is idiomatic for passing errors from child goroutine to parent. We could have separate error/result goroutines, but for the reasons mentioned in my reply to 1., we can get away with just a single channel for now.

Not sure what you're suggesting here with two channels, but that's definitely not what I want. We should at least keep the error handling in `ProcessFromHeight` as it was before, returning an error, and wrap the function that puts `true/false` or `err/nil` into the channel in one spot.

> If the main goroutine exits, all child goroutines will exit, so leaky goroutines aren't an issue in practice.

If the main application exits, all goroutines will exit, but there is no child/parent relationship between goroutines.

> The catch-up routine gets logs via RPC not websocket, so we don't need a persistent connection for correctness.

I'm not worried about the connection. If we want to be able to reconnect, then to do so properly we need to be able to catch up on blocks that were missing while we were disconnected. Currently if we disconnect while still processing historical blocks, we can't start another goroutine because then we either have to have to create a new channel, leaking the old channel, or we have two different goroutines potentially writing conflicting things to the channel.

All this being said, I don't want to block this PR with what I've noticed or the suggestions. I'll create a quick followon PR to this as an example, and we can merge it if it seems worthwhile.